<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>ASEcalculations.R</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: rgb(88, 72, 246)
   }

   pre .number {
     color: rgb(0, 0, 205);
   }

   pre .comment {
     color: rgb(76, 136, 107);
   }

   pre .keyword {
     color: rgb(0, 0, 255);
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: rgb(3, 106, 7);
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>




</head>

<body>
<!-- Automatically generated by RStudio [12861c30b10411e1afa60800200c9a66] -->

<h3>ASEcalculations.R</h3>

<p>johanreimegard &mdash; <em>Sep 9, 2013, 2:51 PM</em></p>

<pre><code class="r"># TODO: Add comment
# 
# Author: johanreimegard

###############################################################################


main &lt;-function(fileName=&quot;BWA_genome.raw.all_FREEC50k.repeatRegions.heterozygousRegions.Crubella_183_only_exons_unique.vcf.Sample.vcf.Rfriendly&quot;, dataDir=&quot;/gulo/proj_nobackup/b2012122/private/ASE&quot;,
                annotation=&quot;BWA_genome.raw.all_FREEC50k.repeatRegions.heterozygousRegions.Crubella_183_only_exons_unique.vcf.Sample.Crubella_183_gene.VCFannotaion&quot;,  
                rounds = 10 ,cutoffNominal=0.005, cutoffAdjusted=0.1){


  file &lt;- paste(dataDir,fileName,sep=&quot;/&quot;)
  annotationFile &lt;- paste(dataDir,annotation,sep=&quot;/&quot;)

  #load dataset  
  print(&quot;Reading data file&quot;)
  sampleData &lt;- read.table(file,header=TRUE,sep =&quot;\t&quot;)
  print(&quot;Reading annotation file&quot;)
  annotationData &lt;- read.table(annotationFile,header=FALSE,sep =&quot;\t&quot;)
  q$annotation=annotationData$V3

  # read in different samples
  DNAInterSamples &lt;- c(&quot;Inter3.1&quot;,&quot;Inter4.1&quot;,&quot;Inter5.1&quot;)
  DNAIntraSamples &lt;- c(&quot;Intra6.3&quot;,&quot;Intra7.2&quot;,&quot;Intra8.2&quot;)
  RNAInterSamples &lt;- c(&quot;Inter3_1_1_F&quot;,&quot;Inter3_1_1_L&quot;,&quot;Inter4_1_1_F&quot;,&quot;Inter4_1_1_L&quot;,&quot;Inter4_1_2_F&quot;,&quot;Inter4_1_2_L&quot;,&quot;Inter4_1_3_F&quot;,&quot;Inter4_1_4_L&quot;,&quot;Inter5_1_1_F&quot;,&quot;Inter5_1_1_L&quot;)
  RNAIntraSamples &lt;- c(&quot;Intra6_3_F&quot;,&quot;Intra6_3_L&quot;,&quot;Intra7_2_1_F&quot;,&quot;Intra7_2_1_L&quot;,&quot;Intra7_2_2_F&quot;,&quot;Intra7_2_2_L&quot;,&quot;Intra7_2_3_F&quot;,&quot;Intra7_2_3_L&quot;,&quot;Intra8_2_1_F&quot;,&quot;Intra8_2_1_L&quot;)
  AllSamples &lt;- c(DNAInterSamples,DNAIntraSamples,RNAInterSamples,RNAIntraSamples)




  # 
  print(&quot;Checking for ASE&quot;)

  ASEinfo = data.frame()
  Inter3_1_DNA &lt;- c(&quot;Inter3.1&quot;)
  Inter3_1_RNA &lt;-c(&quot;Inter3_1_1_F&quot;,&quot;Inter3_1_1_L&quot;)  
  Inter3_1_SampleData &lt;- getPvalues(sampleData,Inter3_1_DNA,Inter3_1_RNA)
  print(Inter3_1_SampleData$Inter3.1_Count1,Inter3_1_SampleData$Inter3.1_Count2)
  ASEinfo &lt;- rbind(ASEinfo, getASEinfo(Inter3_1_SampleData,Inter3_1_RNA,rounds,cutoffNominal,cutoffAdjusted))

  Inter4_1_DNA &lt;- c(&quot;Inter4.1&quot;)
  Inter4_1_RNA &lt;-c(&quot;Inter4_1_1_F&quot;,&quot;Inter4_1_1_L&quot;,&quot;Inter4_1_2_F&quot;,&quot;Inter4_1_2_L&quot;,&quot;Inter4_1_3_F&quot;,&quot;Inter4_1_4_L&quot;)  
  Inter4_1_RNA_Flower &lt;-c(&quot;Inter4_1_1_F&quot;,&quot;Inter4_1_2_F&quot;,&quot;Inter4_1_3_F&quot;)
  Inter4_1_RNA_Leafs &lt;-c(&quot;Inter4_1_1_L&quot;,&quot;Inter4_1_2_L&quot;,&quot;Inter4_1_4_L&quot;)

  Inter4_1_SampleDataFlowers &lt;- getPvalues(sampleData,Inter4_1_DNA,Inter4_1_RNA_Flower)
  nrOfheterozygousGenesExpressedInIntraFlowers = length(unique(Inter4_1_SampleDataFlowers$annotation))
  InterUnionGenesFlowers &lt;- getASEgenesUnion(Inter4_1_SampleDataFlowers,Inter4_1_RNA_Flower,cutoffAdjusted=0.005)
  InterIntersectGenesFlowers &lt;- getASEgenesIntersect(Inter4_1_SampleDataFlowers,Inter4_1_RNA_Flower,cutoffAdjusted=0.005)


  Inter4_1_SampleDataLeafs &lt;- getPvalues(sampleData,Inter4_1_DNA,Inter4_1_RNA_Leafs)
  nrOfheterozygousGenesExpressedInIntraLeafs = length(unique(Inter4_1_SampleDataLeafs$annotation))
  InterUnionGenesLeafs &lt;- getASEgenesUnion(Inter4_1_SampleDataLeafs,Inter4_1_RNA_Leafs,cutoffAdjusted=0.005)
  InterIntersectGenesLeafs &lt;- getASEgenesIntersect(Inter4_1_SampleDataLeafs,Inter4_1_RNA_Leafs,cutoffAdjusted=0.005)



  Both &lt;- intersect(unique(Inter4_1_SampleDataFlowers$annotation),unique(Inter4_1_SampleDataLeafs$annotation))
  FlowerUnique &lt;- detDiff(unique(Inter4_1_SampleDataFlowers$annotation),Both)
  nrOfheterozygousGenesExpressedInIntra = length(Both)
  BothASEUnion = intersect(UnionGenesFlowers,UnionGenesLeafs)
  BothASEIntersect = intersect(IntersectGenesFlowers,IntersectGenesLeafs)






  Inter4_1_SampleData &lt;- getPvalues(sampleData,Inter4_1_DNA,Inter4_1_RNA)
  ASEinfo &lt;- rbind(ASEinfo, getASEinfo(Inter4_1_SampleData,Inter4_1_RNA,rounds,cutoffNominal,cutoffAdjusted))

  Inter5_1_DNA &lt;- c(&quot;Inter5.1&quot;)
  Inter5_1_RNA &lt;-c(&quot;Inter5_1_1_F&quot;,&quot;Inter5_1_1_L&quot;)  
  Inter5_1_SampleData &lt;- getPvalues(sampleData,Inter5_1_DNA,Inter5_1_RNA)
  ASEinfo &lt;- rbind(ASEinfo, getASEinfo(Inter5_1_SampleData,Inter5_1_RNA,rounds,cutoffNominal,cutoffAdjusted))

  Intra6_3_DNA &lt;- c(&quot;Intra6.3&quot;)
  Intra6_3_RNA &lt;-c(&quot;Intra6_3_F&quot;,&quot;Intra6_3_L&quot;)
  Intra6_3_SampleData &lt;- getPvalues(sampleData,Intra6_3_DNA,Intra6_3_RNA)
  ASEinfo &lt;- rbind(ASEinfo, getASEinfo(Intra6_3_SampleData,Intra6_3_RNA,rounds,cutoffNominal,cutoffAdjusted))

  Intra7_2_DNA &lt;- c(&quot;Intra7.2&quot;)
  Intra7_2_RNA_Flower &lt;-c(&quot;Intra7_2_1_F&quot;,&quot;Intra7_2_2_F&quot;,&quot;Intra7_2_3_F&quot;)
  Intra7_2_RNA_Leafs &lt;-c(&quot;Intra7_2_1_L&quot;,&quot;Intra7_2_2_L&quot;,&quot;Intra7_2_3_L&quot;)

  Intra7_2_SampleDataFlower1 &lt;- getPvalues(sampleData,Intra7_2_DNA,Intra7_2_RNA_Flower[1])
  Intra7_2_SampleDataFlower2 &lt;- getPvalues(sampleData,Intra7_2_DNA,Intra7_2_RNA_Flower[2])
  Intra7_2_SampleDataFlower3 &lt;- getPvalues(sampleData,Intra7_2_DNA,Intra7_2_RNA_Flower[3])

  Intra7_2_SampleDataLeafs1 &lt;- getPvalues(sampleData,Intra7_2_DNA,Intra7_2_RNA_Leafs[1])
  Intra7_2_SampleDataLeafs2 &lt;- getPvalues(sampleData,Intra7_2_DNA,Intra7_2_RNA_Leafs[2])
  Intra7_2_SampleDataLeafs3 &lt;- getPvalues(sampleData,Intra7_2_DNA,Intra7_2_RNA_Leafs[3])


  #
  Intra7_2_SampleDataFlowers &lt;- getPvalues(sampleData,Intra7_2_DNA,Intra7_2_RNA_Flower)
  nrOfheterozygousGenesExpressedInIntraFlowers = length(unique(Intra7_2_SampleDataFlowers$annotation))
  IntraUnionGenesFlowers &lt;- getASEgenesUnion(Intra7_2_SampleDataFlowers,Intra7_2_RNA_Flower,cutoffAdjusted=0.005)
  IntraIntersectGenesFlowers &lt;- getASEgenesIntersect(Intra7_2_SampleDataFlowers,Intra7_2_RNA_Flower,cutoffAdjusted=0.005)


  Intra7_2_SampleDataLeafs &lt;- getPvalues(sampleData,Intra7_2_DNA,Intra7_2_RNA_Leafs)
  nrOfheterozygousGenesExpressedInIntraLeafs = length(unique(Intra7_2_SampleDataLeafs$annotation))
  IntraUnionGenesLeafs &lt;- getASEgenesUnion(Intra7_2_SampleDataLeafs,Intra7_2_RNA_Leafs,cutoffAdjusted=0.005)
  IntraIntersectGenesLeafs &lt;- getASEgenesIntersect(Intra7_2_SampleDataLeafs,Intra7_2_RNA_Leafs,cutoffAdjusted=0.005)

  Both &lt;- intersect(unique(Intra7_2_SampleDataFlowers$annotation),unique(Intra7_2_SampleDataLeafs$annotation))
  nrOfheterozygousGenesExpressedInIntra = length(Both)
  BothASEUnion = intersect(UnionGenesFlowers,UnionGenesLeafs)
  BothASEIntersect = intersect(IntersectGenesFlowers,IntersectGenesLeafs)

  InterUnionGenesFlowers
  InterUnionGenesLeafs
  IntraUnionGenesFlowers
  IntraUnionGenesLeafs

  genes &lt;- unique(sampleData$annotation)
  InterFlowersGenes = unique(Inter4_1_SampleDataFlowers$annotation)
  InterLeavesGenes = unique(Inter4_1_SampleDataLeafs$annotation)
  IntraFlowersGenes = unique(Intra7_2_SampleDataFlowers$annotation)
  IntraLeafsGenes = unique(Intra7_2_SampleDataLeafs$annotation)

  InterFlowersGenesIndex = match(InterFlowersGenes,genes)
  InterLeavesGenesIndex  = match(InterLeavesGenes,genes)
  IntraFlowersGenesIndex  = match(IntraFlowersGenes,genes)
  IntraLeafsGenesIndex  = match(IntraLeafsGenes,genes)




  length(setdiff(InterUnionGenesFlowers,InterUnionGenesLeafs))
  length(setdiff(InterUnionGenesLeafs,InterUnionGenesFlowers))
  length(union(InterUnionGenesFlowers,InterUnionGenesLeafs))

  length(setdiff(IntraUnionGenesFlowers,IntraUnionGenesLeafs))
  length(setdiff(IntraUnionGenesLeafs,IntraUnionGenesFlowers))
  length(union(IntraUnionGenesFlowers,IntraUnionGenesLeafs))
    length(intersect(IntraUnionGenesFlowers,IntraUnionGenesLeafs))


  IntraFlowersASE = IntraUnionGenesFlowers
  IntraLeafsASE = IntraUnionGenesLeafs


  ASEinfo &lt;- rbind(ASEinfo, getASEinfo(Intra7_2_SampleData,Intra7_2_RNA,rounds,cutoffNominal,cutoffAdjusted))

  Intra8_2_DNA &lt;- c(&quot;Intra8.2&quot;)
  Intra8_2_RNA &lt;-c(&quot;Intra8_2_1_F&quot;,&quot;Intra8_2_1_L&quot;)
  Intra8_2_SampleData &lt;- getPvalues(sampleData,Intra8_2_DNA,Intra8_2_RNA)
  ASEinfo &lt;- rbind(ASEinfo, getASEinfo(Intra8_2_SampleData,Intra8_2_RNA,rounds,cutoffNominal,cutoffAdjusted))

  #Under development
  #printPvalues(DataSetInterHeteroAbundantBOTHwithPvalue,RNAInterSamples,DNAInterSamples,&quot;test.pdf&quot;)    
  return (ASEinfo)

}


getPvalues &lt;- function(sampleData,DNAsamples,RNAsamples, RNAlow=40, RNAhigh=2000,DNAlow=30,DNAhigh=200){

  # Different filters
  # Only keep heterozygous DNAsamples
  DataSetFiltered &lt;- filterCallData(sampleData, DNAsamples)  
  # Only keep heterozygous samples with DNA total count above 30 in all heterozygous samples
  DataSetFilteredAbundant &lt;- filterTotalData(DataSetFiltered, DNAsamples,c(DNAlow,DNAhigh))  
  # Only keep heterozygous samples with DNA total count above 40 in all heterozygous samples
  DataSetFilteredAbundantBOTH &lt;- filterTotalData(DataSetFilteredAbundant, RNAsamples,c(RNAlow,RNAhigh))  

  # Calculate pValues 
  DataSetFilteredAbundantBOTHwithPvalue &lt;- Pvaluescalculation(DataSetFilteredAbundantBOTH, RNAsamples,DNAsamples)  

  # Correct for multiple testing
  DataSetFilteredAbundantBOTHwithPvalue&lt;-as.data.frame(DataSetFilteredAbundantBOTHwithPvalue)
  DataSetInterHeteroAbundantBOTHwithPvalueCorrected &lt;- MultipleTestCorrection(DataSetFilteredAbundantBOTHwithPvalue,RNAsamples)


  return(DataSetInterHeteroAbundantBOTHwithPvalueCorrected)

}





printPvalues &lt;- function(Dataset,RNAsamples,DNAsamples, pdfFileName){

  pdf(pdfFileName)
  par(mfrow=c(length(RNAsamples),2), bty=&quot;l&quot;, cex=0.6)
  for(i in 1:length(RNAsamples)){
    for(j in 1:length(DNAsamples[j])){
      RNAsample = strsplit(RNAsamples[i], &#39;_&#39;)
      DNAsample = strsplit(DNAsamples[j], &#39;\\.&#39;)
      if( RNAsample[[1]][1]==DNAsample[[1]][1]){

        name = paste(RNAsamples[i], &quot;pValue&quot;, sep=&quot;_&quot;)
        hist(Dataset[name],breaks=40)       
        name = paste(RNAsamples[i], &quot;fraction&quot;, sep=&quot;_&quot;)
        name2 = paste(DNAsamples[j], &quot;fraction&quot;, sep=&quot;_&quot;)
        boxScatterPlot(Dataset[name],Dataset[name],ylabText=&quot;DNA distribution&quot;,xlabText=&quot;RNA distribtuion&quot;,title = paste( &quot;ASE distribution&quot;))  
      }
    }
  }
  dev.off()

}


boxScatterPlot &lt;- function(x,y,ylabText=&quot;Major readCount&quot;,xlabText=&quot;Minor readCount&quot;,title =&quot;Enhanced Scatterplot&quot;){
  par(fig=c(0,0.8,0,0.8), new=TRUE)
  plot(x, y , xlab=xlabText,
       ylab=ylabText)
  par(fig=c(0,0.8,0.55,1), new=TRUE)
  boxplot(x, horizontal=TRUE, axes=FALSE)
  par(fig=c(0.65,1,0,0.8),new=TRUE)
  boxplot(y, axes=FALSE)
  mtext(title, side=3, outer=TRUE, line=-3)

}




filterCallData  &lt;- function(testData , samples, call=&quot;0/1&quot;){

  # filter for heterozygous sites in all samples
  for(i in 1:length(samples)){
    name = paste(samples[i], &quot;Call&quot;, sep=&quot;_&quot;)
    testData &lt;- testData[testData[name]==call,]     
  }
  return (testData)
}

filterTotalData &lt;- function(Dataset , samples,cutoff=c(40,200)){
  for(i in 1:length(samples)){
    name = paste(samples[i], &quot;Total&quot;, sep=&quot;_&quot;)
    Dataset &lt;- Dataset[Dataset[name]&gt;cutoff[1]  ,]      
  }
  return (Dataset)
}


Pvaluescalculation &lt;- function(testData,RNAsamples,DNAsamples){
  for(i in 1:length(RNAsamples)){
    for(j in 1:length(DNAsamples)){
      RNAsample = strsplit(RNAsamples[i], &#39;_&#39;)
      DNAsample = strsplit(DNAsamples[j], &#39;\\.&#39;)
      if( RNAsample[[1]][1]==DNAsample[[1]][1]){
        print(paste(DNAsamples[j], RNAsamples[i], sep = &quot; comparing &quot;))
        count1RNA=paste(RNAsamples[i], &quot;Count1&quot;, sep=&quot;_&quot;)
        TotalRNA=paste(RNAsamples[i], &quot;Total&quot;, sep=&quot;_&quot;)
        count1DNA=paste(DNAsamples[j], &quot;Count1&quot;, sep=&quot;_&quot;)
        TotalDNA=paste(DNAsamples[j], &quot;Total&quot;, sep=&quot;_&quot;)
        pValuesName= paste(RNAsamples[i], &quot;pValue&quot;, sep=&quot;_&quot;)
        #print(testData[count1RNA]/testData[TotalRNA])
        #print(paste(testData[count1DNA][1], testData[TotalDNA][1], sep = &quot; of total &quot;))


        testData[pValuesName]&lt;- apply(testData,1, BinTest,  count1RNA, TotalRNA, count1DNA, TotalDNA)
      }
    }
  }
  return (testData) 

}


MultipleTestCorrection &lt;- function(testData,RNAsamples) {

  for(i in 1:length(RNAsamples)){
    RNAsample = strsplit(RNAsamples[i], &#39;_&#39;)
    pValue=paste(RNAsamples[i], &quot;pValue&quot;, sep=&quot;_&quot;)
    pValue2= as.numeric(testData[,which(colnames(testData)==pValue)])
    pValuesCorrectedName=paste(RNAsamples[i], &quot;pValueCorrected&quot;, sep=&quot;_&quot;)
    testData[pValuesCorrectedName]&lt;- p.adjust(pValue2,method=&quot;BH&quot;)
  }
  return(testData)
}




BinTest &lt;- function(x, count1RNA, TotalRNA,count1DNA,TotalDNA){
  temp &lt;- binom.test(as.numeric(x[count1RNA]),as.numeric(x[TotalRNA]),p=(as.numeric(x[count1DNA])/as.numeric(x[TotalDNA])), alternative=&quot;two.sided&quot;)
  return (temp$p.value)
}


getASEinfo &lt;- function(Dataset,RNAsamples,rounds=10,cutoffNominal=0.005, cutoffAdjusted=0.1){
  info = data.frame()
  for(i in (1:rounds)){
    DatasetOneSNPperGene &lt;- getOneSNPperGene(Dataset) 
    rowInfo &lt;- getASEinformation(DatasetOneSNPperGene,RNAsamples,cutoffNominal, cutoffAdjusted,i)
    info &lt;- rbind(info,rowInfo)
  }  
  return (info)
}




getASEinformation &lt;- function(DatasetOneSNPperGene,RNAsamples,cutoffNominal=0.005, cutoffAdjusted=0.1,round=1){
  info = data.frame()

  for(i in seq(1,length(RNAsamples)-1)){


    columnName  &lt;- paste(RNAsamples[i],&quot;pValueCorrected&quot;,sep=&quot;_&quot;)
    columnName2  &lt;- paste(RNAsamples[i+1],&quot;pValueCorrected&quot;,sep=&quot;_&quot;)

    Flowers &lt;- DatasetOneSNPperGene[which(DatasetOneSNPperGene[columnName] &lt;cutoffAdjusted),]

    Leafs &lt;- DatasetOneSNPperGene[which(DatasetOneSNPperGene[columnName2] &lt;cutoffAdjusted),]
    LeavesList(RNAsamples[i])=Leafs$annotation
    nrOfASEAdjustedFlowers &lt;- length (Flowers[[1]])
    nrOfASEAdjustedLeafs &lt;- length (Leafs[[1]])
    GeneIntersect &lt;- intersect(Flowers$annotation,Leafs$annotation)
    GeneFlower &lt;- setdiff(Flowers$annotation,Leafs$annotation)
    GeneLeafs &lt;- setdiff(Leafs$annotation,Flowers$annotation)
    nrOfASEAdjustedFlowersAndLeaves &lt;- length (GeneIntersect)
    nrOfASEAdjustedFlowersOnly &lt;- length (GeneFlower)
    nrOfASEAdjustedLeafsOnly &lt;- length (GeneLeafs)


    rowinfo &lt;- data.frame(&quot;Name&quot;=RNAsamples[i],&quot;NrOfGenes&quot;=nrOfGenes,&quot;Adjusted_Pvalue&quot;=cutoffAdjusted,&quot;Adjusted_count&quot;=nrOfASEAdjusted)
    info &lt;- rbind(info,rowinfo)

  }
  return (info)
}

getASEgenesUnion &lt;- function(Dataset,RNAsamples,rounds=10,cutoffNominal=0.005, cutoffAdjusted=0.005){
  DatasetOneSNPperGene &lt;- getOneSNPperGene(Dataset) 
  info=NULL
  for(i in 1:length(RNAsamples)){
    rowInfo &lt;- getASEGenes(DatasetOneSNPperGene,RNAsamples[i],cutoffNominal, cutoffAdjusted,i)
    info &lt;- union(info,rowInfo)
  }
  return (info)
}

getASEgenesIntersect &lt;- function(Dataset,RNAsamples,rounds=10,cutoffNominal=0.005, cutoffAdjusted=0.005){
  DatasetOneSNPperGene &lt;- getOneSNPperGene(Dataset) 
  info=NULL
  for(i in 1:length(RNAsamples)){
    rowInfo &lt;- getASEGenes(DatasetOneSNPperGene,RNAsamples[i],cutoffNominal, cutoffAdjusted,i)
    if(i ==1 ){
      info=rowInfo
    }else{
      info &lt;- intersect(info,rowInfo)
    }
  }
  return (info)
}


getASEGenes &lt;- function(DatasetOneSNPperGene,RNAsample,cutoffNominal=0.005, cutoffAdjusted=0.1,round=1){


  columnName  &lt;- paste(RNAsample,&quot;pValueCorrected&quot;,sep=&quot;_&quot;)

  Flowers &lt;- DatasetOneSNPperGene[which(DatasetOneSNPperGene[columnName] &lt;cutoffAdjusted),]

  return(Flowers$annotation)
}

getOneSNPperGene &lt;- function(Dataset,classifier=&quot;annotation&quot;, cutoff = 0.05){
  Dataset$duplicated &lt;- duplicated.random(as.vector(Dataset[,classifier]))
  DataSetOneSNPperGene &lt;- subset(Dataset, duplicated == FALSE)  
  return (DataSetOneSNPperGene)
}

duplicated.random &lt;- function(x, incomparables = FALSE, ...){
  if ( is.vector(x) )
  {
    permutation = sample(length(x))
    x.perm      = x[permutation]
    result.perm = duplicated(x.perm, incomparables, ...)
    result      = result.perm[order(permutation)]
    return(result)
  }
  else if ( is.matrix(x) )
  {
    permutation = sample(nrow(x))
    x.perm      = x[permutation,]
    result.perm = duplicated(x.perm, incomparables, ...)
    result      = result.perm[order(permutation)]
    return(result)
  }
  else
  {
    stop(paste(&quot;duplicated.random() only supports vectors&quot;,
               &quot;matrices for now.&quot;))
  }
}

  plotVenn &lt;- function(A,B,C,D){
  venn.diagram(
    x = list(A,B,C,D),
    filename = &quot;1D-quadruple_Venn.tiff&quot;,
    col = &quot;black&quot;,
    lty = &quot;dotted&quot;,
    lwd = 4,
    fill = c(&quot;cornflowerblue&quot;, &quot;green&quot;, &quot;yellow&quot;, &quot;darkorchid1&quot;),
    alpha = 0.50,
    label.col = c(&quot;orange&quot;, &quot;white&quot;, &quot;darkorchid4&quot;, &quot;white&quot;, &quot;white&quot;, &quot;white&quot;, &quot;white&quot;, &quot;white&quot;, &quot;darkblue&quot;, &quot;white&quot;, &quot;white&quot;, &quot;white&quot;, &quot;white&quot;, &quot;darkgreen&quot;, &quot;white&quot;),
    cex = 2.5,
    fontfamily = &quot;serif&quot;,
    fontface = &quot;bold&quot;,
    cat.col = c(&quot;darkblue&quot;, &quot;darkgreen&quot;, &quot;orange&quot;, &quot;darkorchid4&quot;),
    cat.cex = 2.5,
    cat.fontfamily = &quot;serif&quot;
  );
  }


</code></pre>

</body>

</html>

